src/Features/CSharp/Portable/Snippets/CSharpDoWhileLoopStatementProvider.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageService;
+using Microsoft.CodeAnalysis.Shared.Extensions.ContextQuery;
+using Microsoft.CodeAnalysis.Snippets;
+using Microsoft.CodeAnalysis.Snippets.SnippetProviders;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp.Snippets;
+
+[ExportSnippetProvider(nameof(ISnippetProvider), LanguageNames.CSharp), Shared]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal sealed class CSharpDoWhileLoopStatementProvider() : AbstractConditionalBlockSnippetProvider
+{
+    public override string Identifier => CSharpSnippetIdentifiers.Do;
+
+    public override string Description => CSharpFeaturesResources.do_while_loop;
+
+    protected override SyntaxNode GenerateStatement(SyntaxGenerator generator, SyntaxContext syntaxContext, InlineExpressionInfo? inlineExpressionInfo)
+    {
+        return SyntaxFactory.DoStatement(
+            SyntaxFactory.Block(),
+            (ExpressionSyntax)(inlineExpressionInfo?.Node.WithoutLeadingTrivia() ?? generator.TrueLiteralExpression()));
+    }
+
+    protected override SyntaxNode GetCondition(SyntaxNode node)
+    {
+        var doStatement = (DoStatementSyntax)node;
```

Comment:
nit: int eh future, it would nice to use generics (with teh base type) to make it so that all this is strongly typed.

Reply:
In this particular example this would just move the hard cast one layer above:
https://github.com/dotnet/roslyn/blob/e6c68774586f44370083f0f444a1b165b0ef7182/src/Features/Core/Portable/Snippets/SnippetProviders/AbstractConditionalBlockSnippetProvider.cs#L21-L26
`GetPlaceHolderLocationsList` is from the `AbstractSnippetProvider` class, which is a base class for all snippets. And even if we try to make `AbstractSnippetProvider` generic as well, we will still need to hard cast here (see how `mainChangeNode` is just found via annotation, the cast would be here):
https://github.com/dotnet/roslyn/blob/e6c68774586f44370083f0f444a1b165b0ef7182/src/Features/Core/Portable/Snippets/SnippetProviders/AbstractSnippetProvider.cs#L110-L123

Reply:
Yup.  mkaing these changes in https://github.com/dotnet/roslyn/pull/72892